feat: Allow for the deletion of any document

ksushant6566 opened the pull request:
Resolves #696

Mythie commented on packages/trpc/server/document-router/router.ts:
```diff
@@ -97,15 +99,18 @@ export const documentRouter = router({
         const userId = ctx.user.id;
 
-        return await deleteDraftDocument({ id, userId });
+        if (status === DocumentStatus.DRAFT) {
+          return await deleteDraftDocument({ id, userId });
+        }
+        return await deleteNonDraftDocument({ id, userId, status });
```
Should we simplify this by merging the two and just returning early for draft documents? We typically don't introduce logic in our tRPC routers as doing so makes them less portable for other API types.

ksushant6566 replied:
oh okay, I'll just make a `deleteDocument({ id, userId, status })` function that merges the functionality

Mythie commented on apps/web/src/app/(dashboard)/documents/deleteWithConfirmation.tsx:
```diff
@@ -0,0 +1,26 @@
+import { useState } from 'react';
+
+export const DeleteWithConfirmation = () => {
```
This file doesn't appear to be used? 🤔

ksushant6566 replied:
oops

Mythie commented on packages/email/templates/document-cancel.tsx:
```diff
@@ -0,0 +1,66 @@
+  documentName = 'Open Source Pledge.pdf',
+  assetBaseUrl = 'http://localhost:3002',
+}: DocumentCancelEmailTemplateProps) => {
+  const previewText = `${inviterName} has cancelled the document ${documentName}, you don't need to sign it anymore.`;
```
Feels like a redundant template string since we don't change it based on any conditions? 🤔

ksushant6566 replied:
In my opinion, it just makes it more readable & future-proof. But lmk if you want it changed?

Mythie replied:
Happy to keep, just noting that it doesn't do much 😄 

I try to avoid designing for a possible future and instead design for the today and the potential throw away

Mythie commented on packages/lib/server-only/document/delete-document.ts:
```diff
@@ -0,0 +1,96 @@
+  status: DocumentStatus;
+};
+
+export const deleteDocument = async ({ id, userId, status }: DeleteDocumentOptions) => {
```
Good stuff on making this change 😄

Mythie commented:
I think this is mostly missing interactions with what happens when a signer visits the signing page for a deleted document etc.

If a document was deleted do we redirect them? Show an error page? If we do either of those what do we do for completed documents with a valid token, should we just show the normal completion page?

ksushant6566 commented on apps/web/src/app/(dashboard)/documents/delete-document-dialog.tsx:
```diff
@@ -10,41 +13,46 @@ import {
+  const onDelete = async () => {
     try {
-      await deleteDocument({ id });
+      await deleteDocument({ id, status });
```
yeah you are right!, lemme just make these changes and push quickly

ksushant6566 replied:
wait we are toasting on successful deletes, see OnSuccess of `trpcReact.document.deleteDocument.useMutation`

Mythie replied:
Right I always get mixed up since we've changed things up a little there!

ksushant6566 commented on apps/web/src/app/(dashboard)/documents/delete-document-dialog.tsx:
```diff
@@ -55,6 +63,11 @@ export const DeleteDraftDocumentDialog = ({
     }
   };
 
+  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
```
okay, changing this

ksushant6566 commented on packages/lib/server-only/document/delete-document.ts:
```diff
@@ -0,0 +1,96 @@
+
+  // if the document is pending, send cancellation emails to all recipients
+  if (status === DocumentStatus.PENDING) {
+    const user = await prisma.user.findFirstOrThrow({
+      where: {
+        id: userId,
+      },
+    });
+
+    const document = await prisma.document.findUnique({
+      where: {
+        id: id,
+        userId,
+      },
+      include: {
+        Recipient: true,
+        documentMeta: true,
+      },
+    });
```
I mean we can do that, but maybe this seems a bit more readable?

Mythie replied:
This will involve two round trips to the database while there's potential for the one above to do a join, it's negligible in terms of performance but could add up later on. Happy to leave for now for the sake of readability though 😄

ksushant6566 commented on packages/lib/server-only/document/delete-document.ts:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    if (document.Recipient.length === 0) {
+      throw new Error('Document has no recipients');
```
oh yes, if there are no recipients then we don't need to do anything

Mythie replied:
That's what I was thinking :)

ksushant6566 commented:
> I think this is mostly missing interactions with what happens when a signer visits the signing page for a deleted document etc.
> 
> If a document was deleted do we redirect them? Show an error page? If we do either of those what do we do for completed documents with a valid token, should we just show the normal completion page?

I think for completed documents we can just show the normal page & for pending documents we can disable actions and show a message that **_this document is no longer available to sign_**

wdyt?

Mythie commented:
I'll do a round of UI review with this now as well, I've only been stinging you on code so far 👀

Mythie commented:
This will also need some guards for recipients adding signatures when a document has since been deleted 😄

ksushant6566 commented:
> This will also need some guards for recipients adding signatures when a document has since been deleted 😄

Done, added a check in `sign-field-with-token.ts` .

ksushant6566 commented:
Here is a loom video of the solution :)

https://www.loom.com/share/8f28b9ad4bb64a70b33da9d59fd2a95b?sid=3da77147-be41-486a-af86-cc8fd09afce2

Mythie commented on apps/web/src/app/(dashboard)/documents/delete-document-dialog.tsx:
```diff
@@ -78,7 +102,13 @@ export const DeleteDraftDocumentDialog = ({
 
-            <Button type="button" loading={isLoading} onClick={onDraftDelete} className="flex-1">
+            <Button
+              type="button"
```
This should be a `variant="destructive"` so we know we're nuking something

Mythie commented on apps/web/src/app/(signing)/sign/[token]/complete/page.tsx:
```diff
@@ -92,11 +99,18 @@ export default async function CompletedSigningPage({
-              You will receive an Email copy of the signed document once everyone has signed.
-            </p>
-          ))}
+          .otherwise(() =>
```
Same thing

ksushant6566 commented:
Yo, I believe everything works now, but still do confirm if the logic below makes sense:

What documents users can see:

**own documents**
_All documents created by me, but not the ones I deleted._

**received documents**

Pending documents: 

- not deleted: Yes
- deleted but signed: Yes
- deleted, not signed (Inbox):  No

Completed documents: 

- not deleted: Yes
- deleted: Yes

Note: Only owners can delete documents, users can't delete received documents. That will involve lot more work. (Maybe open another issue?)

PS: thanks a lot for your help throughout 🙌

Mythie commented:
This all sounds correct to me, going to run through a few rounds of testing before I hit the merge button :)
